### PR TITLE
Don't sanitize filename on Unix when `--no-windows-filenames`

### LINKED ISSUE #4547: Unicode "|" character is being replaced in filenames on filesystems that support it
### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.07.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Per the documentation ```--no-restrict-filenames``` should support Unicode characters. However the `|` character, which is valid Unicode is being replaced with `_` in file paths -

## Actual behavior - 

```
$ yt-dlp -vU \
--get-filename \
-o "/%(uploader)s/S01E%(video_autonumber)02d - %(title)s/S01E%(video_autonumber)02d - %(title)s [%(id)s].%(ext)s" \
l4gGWufoIYI

[debug] Command-line config: ['-vU', '--get-filename', '-o', '/%(uploader)s/S01E%(video_autonumber)02d - %(title)s/S01E%(video_autonumber)02d - %(title)s [%(id)s].%(ext)s', 'l4gGWufoIYI']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05e] (zip)
[debug] Python 3.10.4 (CPython 64bit) - Linux-5.17.15-76051715-generic-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: avconv -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: avprobe -bsfs
[debug] exe versions: none
[debug] Optional libraries: brotli-1.0.9, certifi-2020.06.20, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.07.18, Current version: 2022.07.18
yt-dlp is up to date (2022.07.18)
[debug] [youtube] Extracting URL: l4gGWufoIYI
[youtube] l4gGWufoIYI: Downloading webpage
[youtube] l4gGWufoIYI: Downloading android player API JSON
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] l4gGWufoIYI: Downloading 1 format(s): 248+251

/Action BOX/S01E01 - Homemade Plastic Injection Machine _ DIY/S01E01 - Homemade Plastic Injection Machine _ DIY [l4gGWufoIYI].webm
```

## Expected behavior - 
The video title on Youtube is "Homemade Plastic Injection Machine | DIY" so the expected file path is -
```
/Action BOX/S01E01 - Homemade Plastic Injection Machine | DIY/S01E01 - Homemade Plastic Injection Machine | DIY [l4gGWufoIYI].webm
```

## Notes - 
I'm running Ubuntu 21 with a `ext4` filesystem, `|` is a supported character in filenames - 
```
$ touch 'this is a pipe | in a filename'
$ ls
'this is a pipe | in a filename'
```
To the best of my knowledge NTFS (Windows) is the only filesystem that doesn't support `|` in a path. `yt-dlp` already handles filenames differently when running on Windows via the default `--no-windows-filenames`. Removing the `|` restriction for non-NTFS shouldn't cause an issue. 

## Extras - 
While this ticket is specific to the '|' character NTFS is the only filesystem that doesn't support all Unicode characters - 
```
$ touch 'this is a happy filename 😀'
$ ls
'this is a happy filename 😀'
```
It may be worth exploring supporting the entire Unicode character set in `yt-dlp`.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', '--get-filename', '-o', '/%(uploader)s/S01E%(video_autonumber)02d - %(title)s/S01E%(video_autonumber)02d - %(title)s [%(id)s].%(ext)s', 'l4gGWufoIYI']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05e] (zip)
[debug] Python 3.10.4 (CPython 64bit) - Linux-5.17.15-76051715-generic-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: avconv -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: avprobe -bsfs
[debug] exe versions: none
[debug] Optional libraries: brotli-1.0.9, certifi-2020.06.20, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.07.18, Current version: 2022.07.18
yt-dlp is up to date (2022.07.18)
[debug] [youtube] Extracting URL: l4gGWufoIYI
[youtube] l4gGWufoIYI: Downloading webpage
[youtube] l4gGWufoIYI: Downloading android player API JSON
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] l4gGWufoIYI: Downloading 1 format(s): 248+251

/Action BOX/S01E01 - Homemade Plastic Injection Machine _ DIY/S01E01 - Homemade Plastic Injection Machine _ DIY [l4gGWufoIYI].webm
```

### PR DESCRIPTION
**Pull request by pukkandan**
Remove all filename sanitization on non-windows systems except for `/` and `\0` when `--no-windows-filename` is passed. Closes #4547, closes #8464

### CODE REVIEW

File: yt_dlp/YoutubeDL.py [hunk 1]
```diff
@@ -1292,14 +1293,23 @@ def get_value(mdict):
...
-        sanitizer = sanitize if callable(sanitize) else filename_sanitizer
-        sanitize = bool(sanitize)
+        if callable(sanitize):
+            self.deprecation_warning('Passing a callable "sanitize" to YoutubeDL.prepare_outtmpl is deprecated')
+        elif not sanitize:
+            pass
+        elif (sys.platform != 'win32' and not self.params.get('restrictfilenames')
+                and self.params.get('windowsfilenames') is False):
+            def sanitize(key, value):
+                return value.replace('/', '\u29F8').replace('\0', '')
```
Comment (Grub4K) [inline yt_dlp/YoutubeDL.py:1309]: How about
```suggestion
                return value.translate(str.maketrans('/', '\u29F8', '\0'))
```

**Review by Grub4K — APPROVED** [review]

File: yt_dlp/YoutubeDL.py (hunk 1 above)
Comment (pukkandan) [reply yt_dlp/YoutubeDL.py:1309]: Either's fine by me

**Comment by pukkandan** [comment]
While EXT4 supports it, could someone check OS/Python doesn't freak out if filename has control characters or `\r\n`?

**Comment by Grub4K** [comment]
```shell
$ printf 'test\n' > $'\r\n.txt'
$ ls
Permissions Size User   Date Modified       Name
.rw-r--r--     5 grub4k 2024-04-01 12:20:48 \r\n.txt
$ cat $'\r\n.txt'
───────┬──────────────
       │ File:
.txt
───────┼──────────────
   1   │ test
───────┴──────────────
```

File: yt_dlp/YoutubeDL.py (hunk 1 above)
Comment (Grub4K) [reply yt_dlp/YoutubeDL.py:1309]: Since it's short I guess old one is good

**Comment by pukkandan** [comment]
> ```shell
> $ printf 'test\n' > $'\r\n.txt'
> ```

I was more concerned Python's `open` may cause issues, but I've tested and it works

**Comment by Anuskuss** [comment]
Please merge :pray:

**Comment by DavidPesticcio** [comment]
For anyone passing by:
- https://stackoverflow.com/a/38951058/1506858

> "There is no encoding of filenames in Linux (in ext family of filesystems at any rate). Filenames are sequences of bytes, not characters. It is up to application programs to interpret these bytes as UTF-8 or anything else. The filesystem doesn't care."

> "...byte values 0x00 and 0x2F (and no others) are significant to the kernel in pathnames, and cannot appear in a pathname component."

Also:
- - https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html

**Pull request merged**